### DISCUSSION TITLE
* Failed to parse /home/ansible/inventory-dyn-datacenter.yaml with auto plugin:

### DISCUSSION THREAD

**Discussion by Akasurde** [netbox-community/ansible_modules#1225] [General]
### Summary

I did an update on Netbox (3.6.9 to 3.7.0). After this update, Ansible and Netbox cannot link together, I checked the 443 and SQL port and everything its ok but when I try to link both I've got the issue in tittle.

### Issue Type

Bug Report

### Component Name

netbox ansible link issue

### Ansible Version

```console
$ ansible --version
ansible 2.9.6
  config file = /root/.ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.10 (default, Nov 22 2023, 10:22:35) [GCC 9.4.0]
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
DEFAULT_FORKS(/root/.ansible.cfg) = 10
DEFAULT_LOAD_CALLBACK_PLUGINS(/root/.ansible.cfg) = True
DEFAULT_LOG_PATH(/root/.ansible.cfg) = /var/log/ansible.log
DEFAULT_STDOUT_CALLBACK(/root/.ansible.cfg) = yaml
DEPRECATION_WARNINGS(/root/.ansible.cfg) = False
DISPLAY_SKIPPED_HOSTS(/root/.ansible.cfg) = False
HOST_KEY_CHECKING(/root/.ansible.cfg) = False
LOCALHOST_WARNING(/root/.ansible.cfg) = False
PERSISTENT_COMMAND_TIMEOUT(/root/.ansible.cfg) = 60
PERSISTENT_CONNECT_TIMEOUT(/root/.ansible.cfg) = 60
```

### OS / Environment

Ubuntu 20.04.2 LTS

### Steps to Reproduce

```yaml (paste below)
plugin: netbox.netbox.nb_inventory
api_endpoint: https://smssnetbox.mss.lan
token: 563d1ab22xxxxxxxxxxxxxxxxxxxxx
validate_certs: False
config_context: True
```

### Expected Results

I expected that Ansible access to Netbox inventory and parsed it for my script

### Actual Results

```console
* Failed to parse /home/ansible/inventory-dyn-datacenter.yaml with auto plugin:
Failed to parse /home/ansible/inventory-dyn-datacenter.yaml with yaml plugin: Plugin configuration YAML file, not YAML inventory
[WARNING]:  * Failed to parse /home/ansible/inventory-dyn-datacenter.yaml with ini plugin: Invalid host pattern '---' supplied, '---' is normally a sign this is a YAML file.
[WARNING]: Unable to parse /home/ansible/inventory-dyn-datacenter.yaml as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
{
    "_meta": {
        "hostvars": {}
    },
    "all": {
        "children": [
            "ungrouped"
        ]
    }
}
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by Akasurde** [comment]
cc @Sn0w2776

**Comment by Akasurde** [comment]
Original tracelog - https://github.com/ansible/ansible/issues/82636#issuecomment-1966689899

**Comment by Akasurde** [comment]
Hello, I think my Netbox version is 3.7.2 as you can see on screenshot :
![image](https://github.com/ansible/ansible/assets/158482202/3c12c2b5-40b3-42a7-9c6d-b1539696f57b)

And no I can't access to https://smssnetbox.mss.lan/api/docs/?format=openapi?

![image](https://github.com/ansible/ansible/assets/158482202/1d60e7c4-f48d-48e3-b2da-1d042dd25436)

Thanks for help !

**Comment by Sn0w2776** [comment]
@Akasurde what are your guidelines or advice?

**Comment by sc68cal** [comment]
You state that you did an update to the NetBox server and now the inventory plugin does not work. 

Please check your version of the NetBox Ansible collection and the version of the NetBox server API that it supports

**Comment by Akasurde** [comment]
@Sn0w2776 I think Netbox has different product versions. The one I am using is a trial, you might be using the production. Someone from this collection might have a better idea as to why the inventory failing to fetch `https://smssnetbox.mss.lan/api/docs/?format=openapi` in your case. 

I will provide addition information if required. Thanks.

**Comment by sc68cal** [comment]
Another common issue that we see for this kind of issue is when someone has placed a proxy in front of the NetBox server that does not allow access to the required URLs

**Comment by Akasurde** [comment]
@sc68cal The issue is migrated from `ansible/ansible` repo to here. @Sn0w2776 is the original reporter. Sorry for the inconvenience.

**Comment by sc68cal** [comment]
We also had a refactoring of the logic around the fetching of the openapi and caching, which I would recommend updating your collection in order to get the benefit of

**Comment by tweippert** [comment]
I got a similar or maybe the same  issue with Netbox 3.7.3 (3.7.0 also) and netbox.netbox collection 3.17.0:

```
[WARNING]:  * Failed to parse /home/weiti/DN42/automate/inventory/dn42.yml with auto plugin: '/api/dcim/devices/'
  File "/usr/lib/python3.11/site-packages/ansible/inventory/manager.py", line 293, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/usr/lib/python3.11/site-packages/ansible/plugins/inventory/auto.py", line 59, in parse
    plugin.parse(inventory, loader, path, cache=cache)
  File "/home/weiti/.ansible/collections/ansible_collections/netbox/netbox/plugins/inventory/nb_inventory.py", line 2079, in parse
    self.main()
  File "/home/weiti/.ansible/collections/ansible_collections/netbox/netbox/plugins/inventory/nb_inventory.py", line 1926, in main
    self.fetch_api_docs()
  File "/home/weiti/.ansible/collections/ansible_collections/netbox/netbox/plugins/inventory/nb_inventory.py", line 1546, in fetch_api_docs
    for p in  openapi["paths"]["/api/dcim/devices/"]["get"]["parameters"]
```

Testing it against an old netbox installation 2.10.4 it works as expected.

The used inventory file:
```
plugin: netbox.netbox.nb_inventory
api_endpoint: http://xxx..xxx.xxx/netbox
token: <correct token>
validate_certs: False
config_context: False
group_by:
  - device_roles
query_filters:
  - role: router
device_query_filters:
  - has_primary_ip: 'false'
```

**Comment by tweippert** [comment]
FOr my issues it seems there was an problem with an netbox installation within a subdir (/netbox). but the nb_inventory search the query paramters in /api/dcim/devices not in /netbox/api/dcim/devices.  If i add the subdir to the device and vm query parameters all is working.

Is there a config knob to inform about a subdir installation?

**Reply by sc68cal** [reply]
> Is there a config knob to inform about a subdir installation?

No there is not

**Comment by Sn0w2776** [comment]
@tweippert 
For me I think ansible search in this but I don't know why I have this :

[WARNING]:  * Failed to parse /home/ansible/inventory-dyn-datacenter.yaml with auto plugin:      <!DOCTYPE html> <html   lang="en"   data-netbox-url-name=""   data-netbox-base-
path=""               data-netbox-color-mode="unset"           >   <head>     <meta charset="UTF-8" />     <meta       name="viewport"       content="initial-scale=1, maximum-
scale=1, user-scalable=no, width=device-width, viewport-fit=cover"     />           <title>Page Not Found | NetBox</title>      <script       type="text/javascript"
src="/static/setmode.js"       onerror="window.location='/media-failure/?filename=setmode.js'">     </script>      <script type="text/javascript">       (function () {
initMode()       })();       window.CSRF_TOKEN = "Q61sLabrE6l0CWDR7v0E37EyU0iqeWNrNqTIlTddroMNK2WqPLj98lq5KlEYAu1H";     </script>           <link       rel="stylesheet"
href="/static/netbox-external.css?v=3.7.2"       onerror="window.location='/media-failure/?filename=netbox-external.css'"     />     <link       rel="stylesheet"
href="/static/netbox-light.css?v=3.7.2"       onerror="window.location='/media-failure/?filename=netbox-light.css'"     />     <link       rel="stylesheet"
href="/static/netbox-dark.css?v=3.7.2"       onerror="window.location='/media-failure/?filename=netbox-dark.css'"     />     <link       rel="stylesheet"       media="print"
href="/static/netbox-print.css?v=3.7.2"       onerror="window.location='/media-failure/?filename=netbox-print.css'"     />     <link rel="icon" type="image/png"
href="/static/netbox.ico" />     <link rel="apple-touch-icon" type="image/png" href="/static/netbox_touch-icon-180.png" />           <script       type="text/javascript"
src="/static/netbox.js?v=3.7.2"       onerror="window.location='/media-failure/?filename=netbox.js'">     </script>              </head>    <body>     <script
type="text/javascript">       function checkSideNav() {         // Check localStorage to see if the sidebar should be pinned.         var sideNavRaw =
localStorage.getItem('netbox-sidenav');         // Determine if the device has a small screeen. This media query is equivalent to         // bootstrap's media-breakpoint-
down(lg) breakpoint mixin, which is what the sidenav's         // CSS uses.         var isSmallScreen = window.matchMedia('(max-width: 991.98px)').matches;         if (typeof
sideNavRaw === 'string') {           var sideNavState = JSON.parse(sideNavRaw);           if (sideNavState.pinned === true && !isSmallScreen) {             // If the sidebar
should be pinned and this is not a small screen, set the appropriate             // body attributes prior to the rest of the content rendering. This prevents             //
jumpy/glitchy behavior on page reloads.             document.body.setAttribute('data-sidenav-pinned', '');             document.body.setAttribute('data-sidenav-show', '');
document.body.removeAttribute('data-sidenav-hidden');           } else {             document.body.removeAttribute('data-sidenav-pinned');
document.body.setAttribute('data-sidenav-hidden', '');           }         }       }       window.addEventListener('resize', function(){ checkSideNav() });
checkSideNav();     </script>               <div class="container-fluid px-0">     <main class="layout">              <nav class="sidenav noprint" id="sidenav" data-simplebar>
<div class="sidenav-header">               <a class="sidenav-brand" href="/">       <img src="/static/netbox_logo.svg" height="48" class="sidenav-brand-img" alt="NetBox Logo">
</a>           <a class="sidenav-brand-icon" href="/">       <img src="/static/netbox_icon.svg" height="32" class="sidenav-brand-img" alt="NetBox Logo">     </a>
<button class="btn btn-sm btn-ghost-primary sidenav-toggle">       <div class="sidenav-toggle-icon">         <i class="mdi mdi-pin"></i>       </div>     </button>    </div>
<div class="sidenav-inner h-100 mb-auto">           <div class="collapse sidenav-collapse pb-4">                 <ul class="navbar-nav">    </ul>       </div>           <div
class="profile-button-container">           <div class="btn-group">     <a class="btn btn-primary ws-nowrap" type="button" href="/login/?next=/api/docs/">       <i class="mdi
mdi-login-variant"></i> Log In     </a>     <button type="button" class="btn btn-primary dropdown-toggle dropdown-toggle-split" data-bs-toggle="dropdown">       <span
class="visually-hidden">Toggle Dropdown</span>     </button>     <ul class="dropdown-menu dropdown-menu-end">       <li>         <button class="dropdown-item color-mode-
toggle">           <i class="color-mode-icon mdi mdi-lightbulb"></i>           <span class="color-mode-text">Dark Mode</span>         </button>       </li>     </ul>   </div>
</div>   </div> </nav>                <div class="content-container" tabindex="-2">                   <div class="p-2 printonly">           <img src="/static/netbox_logo.svg"
alt="NetBox logo" width="200px" />         </div>                   <nav class="navbar navbar-light sticky-top flex-md-nowrap p-1 mb-3 search container-fluid noprint">
<div class="nav-mobile">               <div class="nav-mobile-top">                 <a class="sidebar-logo p-2 d-block" href="/">                   <img
src="/static/netbox_logo.svg" alt="NetBox logo" height="40" />                 </a>                 <button type="button" aria-label="Toggle Navigation" class="navbar-toggler
sidenav-toggle-mobile">                   <span class="navbar-toggler-icon"></span>                 </button>               </div>               <div class="d-flex my-1 flex-
grow-1 justify-content-center w-100">                 <form class="input-group" action="/search/" method="get">   <input name="q" type="text" aria-label="Search"
placeholder="Search" class="form-control" />   <button class="btn btn-primary" type="submit">     <i class="mdi mdi-magnify"></i>   </button> </form>                </div>
</div>                           <div class="row search-container">                               <div class="col-3 d-flex flex-grow-1 ps-0"></div>
<div class="col-6 d-flex flex-grow-1 justify-content-center">                 <form class="input-group" action="/search/" method="get">   <input name="q" type="text" aria-
label="Search" placeholder="Search" class="form-control" />   <button class="btn btn-primary" type="submit">     <i class="mdi mdi-magnify"></i>   </button> </form>
</div>                               <div class="col-3 d-flex flex-grow-1 pe-0 justify-content-end">                     <div class="btn-group">     <a class="btn btn-primary
ws-nowrap" type="button" href="/login/?next=/api/docs/">       <i class="mdi mdi-login-variant"></i> Log In     </a>     <button type="button" class="btn btn-primary dropdown-
toggle dropdown-toggle-split" data-bs-toggle="dropdown">       <span class="visually-hidden">Toggle Dropdown</span>     </button>     <ul class="dropdown-menu dropdown-menu-
end">       <li>         <button class="dropdown-item color-mode-toggle">           <i class="color-mode-icon mdi mdi-lightbulb"></i>           <span class="color-mode-
text">Dark Mode</span>         </button>       </li>     </ul>   </div>                 </div>              </div>          </nav>                     <div class="text-center
mx-3">             NEXTLANE Cloud           </div>                                                           <div class="title-container px-3 pb-3">
<div id="content-title">                              <h1 class="h2 w-100">Page Not Found</h1>                            </div>                                       </div>
<div id="content" class="container-fluid content px-0 m-0">                                   <div class="px-3">                <div class="row my-5">     <div class="col-6
offset-3">         <div class="card">             <h5 class="card-header text-danger">                 /api/docs/             </h5>             <div class="card-body">
The requested page does not exist.              </div>             <div class="card-footer text-end">                 <a href="/" class="btn btn-sm btn-primary">Home Page</a>
</div>         </div>     </div> </div>              </div>                                          </div>                                      <footer class="footer
container-fluid">                        <div class="row align-items-center justify-content-between mx-0">                <div class="col-sm-12 col-md-auto fs-4 noprint">
<nav class="nav justify-content-center justify-content-lg-start">                                                             <a type="button" class="nav-link"
href="/static/docs/" target="_blank">                       <i title="Docs" class="mdi mdi-book-open-variant text-primary" data-bs-placement="top" data-bs-toggle="tooltip"></i>
</a>                                           <a type="button" class="nav-link" href="/api/" target="_blank">                       <i title="REST API" class="mdi mdi-cloud-
braces text-primary" data-bs-placement="top" data-bs-toggle="tooltip"></i>                     </a>                                           <a type="button" class="nav-link"
href="/api/schema/swagger-ui/" target="_blank">                       <i title="REST API documentation" class="mdi mdi-book text-primary" data-bs-placement="top" data-bs-
toggle="tooltip"></i>                     </a>                                                                <a type="button" class="nav-link" href="/graphql/"
target="_blank">                       <i title="GraphQL API" class="mdi mdi-graphql text-primary" data-bs-placement="top" data-bs-toggle="tooltip"></i>
</a>                                                                <a type="button" class="nav-link" href="https://github.com/netbox-community/netbox" target="_blank">
<i title="Source Code" class="mdi mdi-github text-primary" data-bs-placement="top" data-bs-toggle="tooltip"></i>                     </a>
<a type="button" class="nav-link" href="https://netdev.chat/" target="_blank">                       <i title="Community" class="mdi mdi-slack text-primary" data-bs-
placement="top" data-bs-toggle="tooltip"></i>                     </a>                                    </nav>               </div>                <div class="col-sm-12 col-
md-auto text-center text-lg-end text-muted">                 <span class="d-block d-md-inline"><span title="Feb. 27, 2024 3:32 p.m.">2024-02-27 15:32</span> CET</span>
<span class="ms-md-3 d-block d-md-inline">SmssNETBOX (v3.7.2)</span>               </div>              </div>                    </footer>        </div>      </main>   </div>
<div id="django-messages" class="toast-container">                </div>            <div id="netbox-data" style="display: none!important; visibility: hidden!important">
</div>    </body> </html>

**Comment by tweippert** [comment]
> @tweippert For me I think ansible search in this but I don't know why I have this :
> 
> [WARNING]: * Failed to parse /home/ansible/inventory-dyn-datacenter.yaml with auto plugin: <title>Page Not Found | NetBox</title> <script type="text/javascript" src="/static/setmode.js" onerror="window.location='/media-failure/?filename=setmode.js'"> </script> <script type="text/javascript"> (function () { initMode() })(); window.CSRF_TOKEN = "Q61sLabrE6l0CWDR7v0E37EyU0iqeWNrNqTIlTddroMNK2WqPLj98lq5KlEYAu1H"; </script> <script type="text/javascript" src="/static/netbox.js?v=3.7.2" onerror="window.location='/media-failure/?filename=netbox.js'"> </script> <script type="text/javascript"> function checkSideNav() { // Check localStorage to see if the sidebar should be pinned. var sideNavRaw = localStorage.getItem('netbox-sidenav'); // Determine if the device has a small screeen. This media query is equivalent to // bootstrap's media-breakpoint- down(lg) breakpoint mixin, which is what the sidenav's // CSS uses. var isSmallScreen = window.matchMedia('(max-width: 991.98px)').matches; if (typeof sideNavRaw === 'string') { var sideNavState = JSON.parse(sideNavRaw); if (sideNavState.pinned === true && !isSmallScreen) { // If the sidebar should be pinned and this is not a small screen, set the appropriate // body attributes prior to the rest of the content rendering. This prevents // jumpy/glitchy behavior on page reloads. document.body.setAttribute('data-sidenav-pinned', ''); document.body.setAttribute('data-sidenav-show', ''); document.body.removeAttribute('data-sidenav-hidden'); } else { document.body.removeAttribute('data-sidenav-pinned'); document.body.setAttribute('data-sidenav-hidden', ''); } } } window.addEventListener('resize', function(){ checkSideNav() }); checkSideNav(); </script>
> 
> [ <img alt="NetBox Logo" height="48" src="/static/netbox_logo.svg"> ](/) [ <img alt="NetBox Logo" height="32" src="/static/netbox_icon.svg"> ](/)
> 
> [ Log In ](/login/?next=/api/docs/) Toggle Dropdown
> * Dark Mode
> 
> <img alt="NetBox logo" width="200px" src="/static/netbox_logo.svg">
> [ <img alt="NetBox logo" height="40" src="/static/netbox_logo.svg"> ](/)
> [ Log In ](/login/?next=/api/docs/) Toggle Dropdown
> * Dark Mode
> 
> NEXTLANE Cloud
> # Page Not Found
> ##### /api/docs/
> The requested page does not exist.
> [Home Page](/)
> [ ](/static/docs/) [ ](/api/) [ ](/api/schema/swagger-ui/) [ ](/graphql/) [ ](https://github.com/netbox-community/netbox) [ ](https://netdev.chat/)
> 2024-02-27 15:32 CET SmssNETBOX (v3.7.2)

I think v3.7.2 has moved the api url away from /api/docs/ so the question is, why does your request goes to this url, which version of pynetbox and netbox.netbox collection did you use?
